### PR TITLE
mgr/dashboard: RGW auto refresh is not working

### PR DESCRIPTION
**Pull request by avanthakkar**
mgr/dashboard: RGW auto refresh is not working

Fixes: https://tracker.ceph.com/issues/44667

Signed-off-by: Avan Thakkar <athakkar@redhat.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by votdev** [comment]
I think it was by intention to not reload the data automatically because this action is really expensive (the user and bucket data must be requested for EVERY item via REST API call).

**Comment by avanthakkar** [comment]
> I think it is by intention to not reload the data automatically because this action is really expensive (the user and bucket data must be requested for EVERY item via REST API call).

Okay. So I want to know how we can know if the particular API call will be more expensive compare to others. Because if you will see for osd and pools , there also the list gets reload by requesting osd and pool APIs. So does that cost less compare to users and buckets in object gateway? I just want to know how to decide that?

**Comment by LenzGr** [comment]
We discussed this topic during our backlog grooming call today. The idea is to show a message in the data table if the information has not been refreshed for a certain amount of time, to inform the user it is stale (see the RBD datatable for an existing implementation).

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.html [hunk 1]
```diff
@@ -1,5 +1,4 @@
 <cd-table #table
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.html]: I think we should use an AlertPanelComponent to display the warning to the user, regarding stale data.
```suggestion
<cd-alert-panel *ngIf="TIME CONDITION IS REACHED"
                type="warning"
                i18n>INFORM THE USER ABOUT POSSIBLE STALE DATA</cd-alert-panel>

<cd-table #table
```

**Comment by tspmelo** [comment]
If you can, please rename the PR and commit title to something shorter like `mgr/dashboard: RGW auto refresh is not working`.
And if you need to explain more, use the body of the commit message.
A good indicator that the title is too long, is that Github breaks it into 2 lines :)

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by avanthakkar** [comment]
jenkins test make check

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.html [hunk 2]
```diff
@@ -1,3 +1,6 @@
+<cd-alert-panel *ngIf="isTimeOut"
+                type="warning"
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.html:2]: Maybe use the slim version of the component
```suggestion
                type="warning"
                size="slim"
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.html [hunk 3]
```diff
@@ -1,3 +1,6 @@
...
+                i18n>PLEASE RELOAD THE BUCKET LIST AS THERE IS SOME POSSIBLE STALE DATA</cd-alert-panel>
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.html:3]: Please don't use all caps and change the phrasing.
First inform the user the data might be stale and after inform that he can manually reload.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.ts [hunk 4]
```diff
@@ -34,14 +34,15 @@ export class RgwBucketListComponent {
   columns: CdTableColumn[] = [];
   buckets: object[] = [];
   selection: CdTableSelection = new CdTableSelection();
-
+  isTimeOut = true;
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.ts:37]: This should start as `false` and I think `isStale `is a better var name.
```suggestion
  isStale= false;
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.ts [hunk 5]
```diff
@@ -80,6 +81,17 @@ export class RgwBucketListComponent {
...
+  }
+  timeConditionReached() {
+    this.isTimeOut = !this.isTimeOut;
+    this.ngZone.runOutsideAngular(() => {
+      setTimeout(() => {
+        this.ngZone.run(() => {
+          this.timeConditionReached();
+        });
+      }, 10000);
+    });
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.ts:94]: We should simply set isStale to true.
There is no need to keep running this method.
It should only be called after its initiated and each time the user clicks the refresh.
and should cancel the previous timeout.

```suggestion
    this.timeConditionReached();
  }
  
  timeConditionReached() {
    clearTimeout(staleTimeout);
    this.ngZone.runOutsideAngular(() => {
      staleTimeout = setTimeout(() => {
        this.ngZone.run(() => {
          this.isStale = true;
        });
      }, 10000);
    });
```

Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.ts:94]: what's staleTimeout here?

Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.ts:94]: Forgot to add that, its a number. 
I have updated my suggestion on https://github.com/ceph/ceph/pull/34142/files#diff-d7cceb9140237e8f690e610324376cb7R37
If there are type conflicts with this, change `setTimeout `to `window.setTimeout`

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.html (hunk 2 above)
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.html:2]: And in ngIf need to write "isStale" ,right?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.html (hunk 3 above)
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.html:3]: 'The bucket list data might me stale, can be manually reloaded' is this fine?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.ts (hunk 5 above)
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.ts:94]: By making the changes suggested the warning remains still..so is it expected behavior or it just get refreshed after 10s?

Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.ts:94]: I think it should remain still after the 10s.
Only when the user refreshes or click the close button, it should disappear .

Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.ts:94]: the warning is remaining forever

Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.ts:94]: Sorry, my bad..forget to set isStale=false. It's working now

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.ts [hunk 6]
```diff
@@ -80,9 +83,21 @@ export class RgwBucketListComponent {
...
+      this.staleTimeout = window.setTimeout(() => {
+        this.ngZone.run(() => {
+          this.isStale = true;
+        });
+      }, 10000);
+    });
   }
 
   getBucketList(context: CdTableFetchDataContext) {
+    this.timeConditionReached();
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.ts:100]: You are missing setting `isStale` to `false` when calling `timeConditionReached()`.
You can either put the code here:
```suggestion
    this.isStale = true;
    this.timeConditionReached();
```
or add it in the 1st line of `timeConditionReached()`

Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.ts:100]: I guess, this.isStale=false, right?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.html [hunk 7]
```diff
@@ -1,3 +1,7 @@
+<cd-alert-panel *ngIf="isStale"
+                type="warning"
+                size="slim"
+                i18n>The bucket list data might me stale, can be manually reloaded</cd-alert-panel>
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.html:4]:
```suggestion
                i18n>The bucket list data might be stale. If needed, you can manually reload it.</cd-alert-panel>
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-list/rgw-user-list.component.ts [hunk 8]
```diff
@@ -103,9 +105,22 @@ export class RgwUserListComponent {
       canBePrimary: (selection: CdTableSelection) => selection.hasMultiSelection
     };
     this.tableActions = [addAction, editAction, deleteAction];
+    this.timeConditionReached();
+  }
+  timeConditionReached() {
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-list/rgw-user-list.component.ts:110]:
```suggestion
  }
  
  timeConditionReached() {
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.html (hunk 7 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.html:4]: same for the other component

Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-list/rgw-bucket-list.component.html]: Done!

**Comment by avanthakkar** [comment]
jenkins test make check

**Pull request merged**